pkg/network/node/networkpolicy.go

```diff
@@ -303,8 +304,34 @@ func (np *networkPolicyPlugin) SyncVNIDRules() {
 	}
 }
 
+// Match namespaces against a selector, using a cache so that, eg, when a new Namespace is
+// added, we only figure out if it matches "name: default" once, rather than recomputing
+// the set of namespaces that match that selector for every single "allow-from-default"
+// policy in the cluster.
+//
+// Yes, if a selector matches against multiple labels then the order they appear in
+// cacheKey here is non-deterministic, but that just means that, eg, we might compute the
+// results twice rather than just once, and twice is still better than 10,000 times.
+//
+// (The match cache gets flushed as needed by refreshNetworkPolicies.)
+func (np *networkPolicyPlugin) selectNamespacesInternal(sel labels.Selector) map[string]uint32 {
+	cacheKey := sel.String()
+	namespaces, wasCached := np.nsMatchCache[cacheKey]
+	if !wasCached {
+		namespaces = make(map[string]uint32)
+		for vnid, ns := range np.namespaces {
```

Comment:
Do we need to flush the cache on `np.namespaces` changes too?

Reply:
Hm... VNID changes aren't allowed in ovs-networkpolicy, so we don't have to worry about updates. In between the deletion of a NetNamespace and its Namespace, the cache might temporarily contain stale rules referring to non-existent VNIDs, which isn't optimal, but won't hurt anything. The only way we could get actually *wrong* rules would be if we got:

    - ADD Namespace bar
    - DEL NetNamespace foo, vnid 42
    - (cache may contain stale rules)
    - ADD NetNamespace bar, vnid 42
    - (cache may contain incorrect rules)
    - DEL Namespace foo
    - (cache is flushed)

I don't think an ADD by itself could cause problems in any circumstances.